src/Analyzer/Passes/AggregateFunctionOfGroupByKeysPass.cpp

```diff
@@ -0,0 +1,186 @@
+
+        if (it == group_by_keys.end())
+            keep_aggregator = true;
+    }
+
+    void visitImpl(const QueryTreeNodePtr & node)
+    {
+        if (keep_aggregator)
+            return;
+
+        if (node->as<ListNode>())
+            return;
+
+        if (auto * function_node = node->as<FunctionNode>())
+        {
+            visitFunction(function_node);
+        }
+        else if (auto * column = node->as<ColumnNode>())
+        {
+            visitColumn(column);
+        }
+    }
+
+private :
+    const QueryTreeNodes & group_by_keys;
+    bool & keep_aggregator;
+};
+
+/// Try to eliminate min/max/any/anyLast.
+class EliminateFunctionVisitor : public InDepthQueryTreeVisitorWithContext<EliminateFunctionVisitor>
+{
+public:
+    using Base = InDepthQueryTreeVisitorWithContext<EliminateFunctionVisitor>;
+    using Base::Base;
+
+    using GroupByKeysStack = std::vector<QueryTreeNodes>;
+
+    void enterImpl(QueryTreeNodePtr & node)
+    {
+        if (!getSettings().optimize_aggregators_of_group_by_keys)
+            return;
+
+        /// (1) collect group by keys
+        if (auto * query_node = node->as<QueryNode>())
+        {
+            if (!query_node->hasGroupBy())
+            {
+                group_by_keys_stack.push_back({});
+            }
+            else if (query_node->isGroupByWithTotals() || query_node->isGroupByWithCube() || query_node->isGroupByWithRollup())
+            {
+                /// Keep aggregator if group by is with totals/cube/rollup.
+                group_by_keys_stack.push_back({});
+            }
+            else
+            {
+                QueryTreeNodes group_by_keys;
+                for (auto & group_key : query_node->getGroupBy().getNodes())
+                {
+                    /// for grouping sets case
+                    if (auto * list = group_key->as<ListNode>())
+                    {
+                        for (auto & group_elem : list->getNodes())
+                            group_by_keys.push_back(group_elem);
+                    }
+                    else
+                    {
+                        group_by_keys.push_back(group_key);
+                    }
+                }
+                group_by_keys_stack.push_back(std::move(group_by_keys));
+            }
+        }
+        /// (2) Try to eliminate any/min/max
+        else if (auto * function_node = node->as<FunctionNode>())
+        {
+            if (!function_node
+                || !(function_node->getFunctionName() == "min" || function_node->getFunctionName() == "max"
+                     || function_node->getFunctionName() == "any" || function_node->getFunctionName() == "anyLast"))
+                return;
+
+            if (!function_node->getArguments().getNodes().empty())
+            {
+                bool keep_aggregator = false;
+
+                KeepEliminateFunctionVisitor visitor(group_by_keys_stack.back(), keep_aggregator);
```

Comment:
I'd prefer to implement this check in non-recursive way. The same as we do in `OptimizeGroupByFunctionKeysPass`.